File: ai_fleet/agent/framework.py
"""Workflow-driven deep-agent framework (port of agent/framework.js).

Both the planning and coding agents are the SAME machine configured by a
declarative workflow descriptor (ai_fleet/agent/workflows/<name>.py). A workflow
declares which SKILLS to load, which TOOLS to attach, the backend kind, the
system prompt, and run limits. This module turns that descriptor into a live
`deepagents` agent and runs it.

Backends:
  - 'filesystem' — FilesystemBackend (read/write files, NO shell). Used by the planner.
  - 'shell'      — LocalShellBackend (fs + shell). Used by the coder, rooted at an
    isolated git workspace.
"""

from __future__ import annotations

import importlib
import os
import shutil
import stat
import subprocess
import tempfile
import uuid
from pathlib import Path

from .safe_read import install_safe_read

SKILLS_SRC = Path(__file__).parent / "skills"
SKILLS_DEST_DIRNAME = ".agent-skills"
SKILLS_OWNER_MARKER = ".tech-symphony-managed"
SKILLS_OWNER_MARKER_CONTENT = "tech-symphony-agent-skills-v1\n"


def _lstat_or_none(p):
    try:
        return os.lstat(p)
    except FileNotFoundError:
        return None


def _is_dir(p):
    try:
        return stat.S_ISDIR(os.stat(p).st_mode)
    except OSError:
        return False


def _assert_contained(real_root, candidate):
    relative = os.path.relpath(os.path.abspath(candidate), real_root)
    if relative == ".." or relative.startswith(f"..{os.sep}") or os.path.isabs(relative):
        raise ValueError(f"Refusing to install skills outside destination root: {candidate}")


def _assert_no_symlinks(target):
    st = _lstat_or_none(target)
    if not st:
        return
    if stat.S_ISLNK(st.st_mode):
        raise ValueError(f"Refusing symbolic link inside skill destination: {target}")
    if not stat.S_ISDIR(st.st_mode):
        return
    for entry in os.listdir(target):
        _assert_no_symlinks(os.path.join(target, entry))


def _validate_skill_name(name):
    if (
        not isinstance(name, str)
        or not name
        or name == "."
        or name == ".."
        or os.path.basename(name) != name
        or "/" in name
        or "\\" in name
    ):
        raise ValueError(f"Invalid skill name: {name}")


def _tracked_skill_paths(root):
    try:
        output = subprocess.run(
            ["git", "-C", str(root), "ls-files", "--", SKILLS_DEST_DIRNAME],
            capture_output=True,
            text=True,
            check=True,
        ).stdout
        return [line.strip() for line in output.splitlines() if line.strip()]
    except Exception:
        # Scratch workspaces are intentionally not Git repositories.
        return []


def _claim_skills_directory(dest, real_root):
    existing = _lstat_or_none(dest)
    if existing:
        if stat.S_ISLNK(existing.st_mode):
            raise ValueError(f"Refusing symbolic-link skill destination: {dest}")
        if not stat.S_ISDIR(existing.st_mode):
            raise ValueError(f"Refusing non-directory skill destination: {dest}")
        _assert_contained(real_root, os.path.realpath(dest))
        _assert_no_symlinks(dest)
        marker = os.path.join(dest, SKILLS_OWNER_MARKER)
        marker_stat = _lstat_or_none(marker)
        if not marker_stat or stat.S_ISLNK(marker_stat.st_mode) or not stat.S_ISREG(marker_stat.st_mode):
            raise ValueError(f"Refusing project-owned skill directory without a valid ownership marker: {dest}")
        with open(marker, encoding="utf-8") as fh:
            if fh.read() != SKILLS_OWNER_MARKER_CONTENT:
                raise ValueError(f"Refusing skill directory with an invalid ownership marker: {dest}")
        return

    _assert_contained(real_root, dest)
    os.mkdir(dest)
    _assert_contained(real_root, os.path.realpath(dest))
    fd = os.open(os.path.join(dest, SKILLS_OWNER_MARKER), os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
    with os.fdopen(fd, "w", encoding="utf-8") as fh:
        fh.write(SKILLS_OWNER_MARKER_CONTENT)


def install_skills(dest_root, skill_names=None):
    """Copy named skills into dest_root/.agent-skills/ and return their
    backend-relative paths (e.g. /.agent-skills/software-planning/)."""
    root = os.path.abspath(dest_root)
    root_stat = _lstat_or_none(root)
    if not root_stat or not stat.S_ISDIR(root_stat.st_mode):
        raise ValueError(f"Skill destination root is not a directory: {root}")
    real_root = os.path.realpath(root)
    dest = os.path.join(real_root, SKILLS_DEST_DIRNAME)
    available = [n for n in os.listdir(SKILLS_SRC) if _is_dir(os.path.join(SKILLS_SRC, n))]
    names = skill_names if isinstance(skill_names, list) and skill_names else available
    for name in names:
        _validate_skill_name(name)

    tracked = _tracked_skill_paths(root)
    if tracked:
        raise ValueError(
            f"Refusing to install framework skills over tracked project files in {SKILLS_DEST_DIRNAME}: {', '.join(tracked[:3])}"
        )

    _claim_skills_directory(dest, real_root)
    paths = []
    for name in dict.fromkeys(names):
        from_ = os.path.join(SKILLS_SRC, name)
        if not _is_dir(from_):
            continue
        _assert_no_symlinks(from_)
        to = os.path.join(dest, name)
        _assert_contained(real_root, to)
        _assert_no_symlinks(to)
        shutil.rmtree(to, ignore_errors=True)
        shutil.copytree(from_, to)
        paths.append(f"/{SKILLS_DEST_DIRNAME}/{name}/")
    return paths


def build_backend(kind, root_dir, opts=None):
    """Build the backend for a workflow kind, rooted at root_dir."""
    opts = opts or {}
    from deepagents.backends import FilesystemBackend, LocalShellBackend

    if kind == "shell":
        from .repository_broker import build_safe_agent_env

        env = build_safe_agent_env(opts.get("env") or dict(os.environ), root_dir)
        return LocalShellBackend(
            root_dir=str(root_dir), virtual_mode=False, env=env, inherit_env=False, timeout=opts.get("timeout") or 600
        )
    return FilesystemBackend(root_dir=str(root_dir), virtual_mode=False)


def load_workflow(name):
    """Load a workflow descriptor by name from workflows/<name>.py."""
    try:
        module = importlib.import_module(f"ai_fleet.agent.workflows.{name}")
    except ModuleNotFoundError:
        raise ValueError(f'Unknown workflow "{name}".')
    return module.WORKFLOW


def prepare_scratch(workflow):
    """Prepare an isolated scratch directory for a repo-less workflow (e.g. planner)."""
    root_dir = os.path.join(tempfile.gettempdir(), "techsym-agent", f"{workflow['name']}-{str(uuid.uuid4())[:8]}")
    os.makedirs(root_dir, exist_ok=True)
    skill_paths = install_skills(root_dir, workflow.get("skills"))

    def cleanup():
        shutil.rmtree(root_dir, ignore_errors=True)

    return {"rootDir": root_dir, "skillPaths": skill_paths, "cleanup": cleanup}


def _resolve_system_prompt(workflow, ctx):
    sp = workflow.get("systemPrompt")
    return sp(ctx) if callable(sp) else sp


def build_agent(workflow, llm, backend=None, skill_paths=None, root_dir=None, ctx=None, extra_tools=None, env=None):
    """Build a deep agent from a workflow descriptor."""
    from deepagents import create_deep_agent

    from . import llm as llm_module
    from . import tools as tool_registry
    from .fs_arg_normalizer import create_fs_arg_normalizer_middleware

    ctx = ctx or {}
    extra_tools = extra_tools or []
    skills = skill_paths
    be = backend
    if be is None:
        if not root_dir:
            raise ValueError("build_agent needs a backend or a root_dir.")
        skills = skills if skills is not None else install_skills(root_dir, workflow.get("skills"))
        be = build_backend(workflow.get("backend"), root_dir, {"timeout": workflow.get("shellTimeoutSec"), "env": env})
    be = install_safe_read(be)
    tools = [*tool_registry.build_many(workflow.get("tools"), ctx), *extra_tools]
    system_prompt = _resolve_system_prompt(workflow, ctx)
    middleware = [create_fs_arg_normalizer_middleware()]
    agent = create_deep_agent(
        model=llm_module.create_chat_model(llm),
        backend=be,
        skills=skills,
        tools=tools,
        system_prompt=system_prompt,
        middleware=middleware,
    )
    return {"agent": agent, "backend": be, "skillPaths": skills, "tools": tools}


async def run_workflow(
    workflow,
    llm,
    user_message,
    backend=None,
    skill_paths=None,
    root_dir=None,
    ctx=None,
    invoke_config=None,
    runtime="deepagent",
    workflow_pattern="sequential",
    env=None,
):
    """Run a workflow agent to completion on a single user message and return
    {result, messages, finalText}."""
    from .runtimes import execute_agent_runtime, normalize_agent_runtime, effective_agent_runtime

    ctx = ctx or {}
    invoke_config = invoke_config or {}
    scratch = None
    if backend is None and not root_dir:
        scratch = prepare_scratch(workflow)
        root_dir = scratch["rootDir"]
        skill_paths = scratch["skillPaths"]
    try:
        requested_runtime = normalize_agent_runtime(runtime, strict=True)
        runtime_id = effective_agent_runtime(requested_runtime, llm, strict=True, workflow=workflow.get("name"))
        config = {
            "recursionLimit": workflow.get("recursionLimit") or 24,
            "tags": workflow.get("tags") or [],
            **invoke_config,
        }
        deep_agent_invoke = None
        if runtime_id == "deepagent":
            from .mcp import load_mcp_tools

            extra_tools = await load_mcp_tools(workflow.get("mcp"), ctx)
            built = build_agent(
                workflow, llm, backend=backend, skill_paths=skill_paths, root_dir=root_dir, ctx=ctx, extra_tools=extra_tools, env=env
            )
            agent = built["agent"]

            async def deep_agent_invoke(prompt, traced_config):
                # The runtime wrapper owns the LangSmith root run id; drop it so the
                # nested LangGraph invocation does not create a duplicate run.
                child_config = {k: v for k, v in (traced_config or {}).items() if k != "runId"}
                return await agent.ainvoke({"messages": [{"role": "user", "content": prompt}]}, child_config or None)

        return await execute_agent_runtime(
            {
                "runtime": requested_runtime,
                "workflowPattern": workflow_pattern,
                "prompt": user_message,
                "workflow": workflow.get("name"),
                "llm": llm,
                "rootDir": root_dir,
                "backendKind": workflow.get("backend"),
                "systemPrompt": workflow.get("systemPrompt"),
                "maxTurns": workflow.get("recursionLimit") or 24,
                "ctx": ctx,
                "env": env,
                "invokeConfig": config,
                "tags": workflow.get("tags") or [],
                "deepAgentInvoke": deep_agent_invoke,
                "lastText": last_text,
            }
        )
    finally:
        if scratch:
            scratch["cleanup"]()


def content_to_text(content):
    """Normalize message content (string or content-block list) to plain text."""
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        parts = []
        for c in content:
            if isinstance(c, str):
                parts.append(c)
            elif isinstance(c, dict):
                parts.append(c.get("text") or "")
            else:
                parts.append(getattr(c, "text", "") or "")
        return "".join(parts)
    return ""


def message_text(msg):
    """Text of a chat message, tolerant of reasoning models."""
    if not msg:
        return ""
    content = msg.get("content") if isinstance(msg, dict) else getattr(msg, "content", None)
    primary = content_to_text(content)
    if primary and primary.strip():
        return primary
    ak = (msg.get("additional_kwargs") if isinstance(msg, dict) else getattr(msg, "additional_kwargs", None)) or {}
    return content_to_text(ak.get("reasoning_content") or ak.get("reasoning") or "")


def last_text(result):
    messages = (result or {}).get("messages") if isinstance(result, dict) else getattr(result, "messages", None)
    messages = messages or []
    return message_text(messages[-1]) if messages else ""

File: packages/shared/src/agent/framework.js
'use strict';

const fs = require('fs');
const os = require('os');
const path = require('path');
const crypto = require('crypto');
const { execFileSync } = require('child_process');
const { resolveSkillsSrc } = require('../config');
const toolRegistry = require('./tools');
const { executeAgentRuntime, normalizeAgentRuntime, effectiveAgentRuntime } = require('./runtimes');
const deepAgentHarness = require('./harnesses/deepagent');
const { applyPolicyToWorkflow, skillNameFromPath } = require('./settings-policy');
const { withResources } = require('./trace-annotations');

/**
 * Workflow-driven deep-agent framework.
 *
 * Both the planning and coding agents are the SAME machine configured by a
 * declarative *workflow file* (server/agent/workflows/<name>.workflow.js). A
 * workflow declares which SKILLS to load, which TOOLS to attach, the backend
 * kind, the system prompt, and run limits. This module owns generic workflow
 * orchestration; the DeepAgent harness owns its SDK-specific construction.
 *
 * Backends:
 *   - 'filesystem' — FilesystemBackend (read/write files, NO shell). Used by the
 *     planner: it only needs its skills on disk + the web_search tool, so denying
 *     shell removes an unnecessary capability (ai-prompt-injection: least tools).
 *   - 'shell'      — LocalShellBackend (fs + shell). Used by the coder, rooted at
 *     an isolated git workspace.
 *
 * Workflow shape:
 *   { name, description, backend: 'filesystem'|'shell', skills: string[],
 *     tools: string[], systemPrompt: string | (ctx)=>string,
 *     recursionLimit?: number, tags?: string[], shellTimeoutSec?: number }
 */

const SKILLS_DEST_DIRNAME = '.agent-skills';
const SKILLS_OWNER_MARKER = '.tech-symphony-managed';
const SKILLS_OWNER_MARKER_CONTENT = 'tech-symphony-agent-skills-v1\n';
const WORKFLOWS_DIR = path.join(__dirname, 'workflows');

/**
 * Copy the named skills from the configured skills source into
 * `destRoot/.agent-skills/` and return their backend-relative paths (e.g.
 * `/.agent-skills/software-planning/`). With no names, installs every available
 * skill (back-compat with the coder's previous "install all" behavior).
 *
 * The source directory is resolved per call from config (resolveSkillsSrc): the
 * vendored `skills/` dir by default, or a version-pinned gcsfuse mount subdir
 * (`$SKILLS_ROOT/$SKILLS_VERSION`) in the cloud. See packages/shared/src/config.js.
 */
function installSkills(destRoot, skillNames) {
  const skillsSrc = resolveSkillsSrc();
  const root = path.resolve(destRoot);
  const rootStat = lstatOrNull(root);
  if (!rootStat || !rootStat.isDirectory()) {
    throw new Error(`Skill destination root is not a directory: ${root}`);
  }
  const realRoot = fs.realpathSync(root);
  const dest = path.join(realRoot, SKILLS_DEST_DIRNAME);
  const available = fs.readdirSync(skillsSrc).filter((n) => isDir(path.join(skillsSrc, n)));
  const names = Array.isArray(skillNames) && skillNames.length ? skillNames : available;
  for (const name of names) validateSkillName(name);

  const tracked = trackedSkillPaths(root);
  if (tracked.length) {
    throw new Error(
      `Refusing to install framework skills over tracked project files in ${SKILLS_DEST_DIRNAME}: ${tracked.slice(0, 3).join(', ')}`
    );
  }

  claimSkillsDirectory(dest, realRoot);
  const paths = [];
  for (const name of new Set(names)) {
    const from = path.join(skillsSrc, name);
    if (!isDir(from)) continue; // skip unknown skill names rather than throw
    assertNoSymlinks(from);
    const to = path.join(dest, name);
    assertContained(realRoot, to);
    assertNoSymlinks(to);
    fs.rmSync(to, { recursive: true, force: true });
    fs.cpSync(from, to, { recursive: true });
    paths.push(`/${SKILLS_DEST_DIRNAME}/${name}/`);
  }
  return paths;
}

function validateSkillName(name) {
  if (
    typeof name !== 'string'
    || !name
    || name === '.'
    || name === '..'
    || path.basename(name) !== name
    || name.includes('/')
    || name.includes('\\')
  ) {
    throw new Error(`Invalid skill name: ${String(name)}`);
  }
}

function trackedSkillPaths(root) {
  try {
    const output = execFileSync('git', ['-C', root, 'ls-files', '--', SKILLS_DEST_DIRNAME], {
      encoding: 'utf8',
      stdio: ['ignore', 'pipe', 'ignore'],
    });
    return output.split(/\r?\n/).map((line) => line.trim()).filter(Boolean);
  } catch (_) {
    // Scratch workspaces are intentionally not Git repositories.
    return [];
  }
}

function claimSkillsDirectory(dest, realRoot) {
  const existing = lstatOrNull(dest);
  if (existing) {
    if (existing.isSymbolicLink()) {
      throw new Error(`Refusing symbolic-link skill destination: ${dest}`);
    }
    if (!existing.isDirectory()) {
      throw new Error(`Refusing non-directory skill destination: ${dest}`);
    }
    assertContained(realRoot, fs.realpathSync(dest));
    assertNoSymlinks(dest);
    const marker = path.join(dest, SKILLS_OWNER_MARKER);
    const markerStat = lstatOrNull(marker);
    if (!markerStat || markerStat.isSymbolicLink() || !markerStat.isFile()) {
      throw new Error(`Refusing project-owned skill directory without a valid ownership marker: ${dest}`);
    }
    if (fs.readFileSync(marker, 'utf8') !== SKILLS_OWNER_MARKER_CONTENT) {
      throw new Error(`Refusing skill directory with an invalid ownership marker: ${dest}`);
    }
    return;
  }

  assertContained(realRoot, dest);
  fs.mkdirSync(dest);
  assertContained(realRoot, fs.realpathSync(dest));
  fs.writeFileSync(path.join(dest, SKILLS_OWNER_MARKER), SKILLS_OWNER_MARKER_CONTENT, {
    encoding: 'utf8',
    flag: 'wx',
    mode: 0o600,
  });
}

function assertContained(realRoot, candidate) {
  const relative = path.relative(realRoot, path.resolve(candidate));
  if (relative === '..' || relative.startsWith(`..${path.sep}`) || path.isAbsolute(relative)) {
    throw new Error(`Refusing to install skills outside destination root: ${candidate}`);
  }
}

function assertNoSymlinks(target) {
  const stat = lstatOrNull(target);
  if (!stat) return;
  if (stat.isSymbolicLink()) {
    throw new Error(`Refusing symbolic link inside skill destination: ${target}`);
  }
  if (!stat.isDirectory()) return;
  for (const entry of fs.readdirSync(target)) {
    assertNoSymlinks(path.join(target, entry));
  }
}

function lstatOrNull(p) {
  try {
    return fs.lstatSync(p);
  } catch (error) {
    if (error && error.code === 'ENOENT') return null;
    throw error;
  }
}

function isDir(p) {
  try {
    return fs.statSync(p).isDirectory();
  } catch (_) {
    return false;
  }
}

/** Back-compat wrapper; DeepAgent backend construction is harness-owned. */
function buildBackend(kind, rootDir, opts = {}) {
  return deepAgentHarness.buildBackend(kind, rootDir, opts);
}

/** Load a workflow descriptor by name from workflows/<name>.workflow.js. */
function loadWorkflow(name) {
  const file = path.join(WORKFLOWS_DIR, `${name}.workflow.js`);
  if (!fs.existsSync(file)) {
    throw new Error(`Unknown workflow "${name}" (expected ${file}).`);
  }
  return require(file);
}

/**
 * Prepare an isolated scratch directory for a workflow that has no repo of its
 * own (e.g. the planner). Installs the workflow's skills there so a
 * FilesystemBackend rooted at the dir can load them. Returns a cleanup fn.
 */
function prepareScratch(workflow) {
  const rootDir = path.join(os.tmpdir(), 'techsym-agent', `${workflow.name}-${crypto.randomUUID().slice(0, 8)}`);
  fs.mkdirSync(rootDir, { recursive: true });
  const skillPaths = installSkills(rootDir, workflow.skills);
  const cleanup = () => fs.rmSync(rootDir, { recursive: true, force: true });
  return { rootDir, skillPaths, cleanup };
}

/** Back-compat wrapper; DeepAgent graph/model/middleware wiring is harness-owned. */
function buildAgent(options) {
  return deepAgentHarness.buildAgent(options, { installSkills });
}

/**
 * Resolve the configured (available) skill/tool/plugin NAMES for a run, for
 * trace metadata. Prefer the concrete set a deepagent build resolved
 * (`resolvedTools`/`resolvedSkills`, already policy-narrowed); otherwise fall
 * back to the workflow's policy-applied declaration so non-deepagent runtimes
 * still record what they were configured with. Plugins are the workflow's MCP
 * servers. Names only — never args or secrets.
 */
function configuredResourceNames({ workflow, effective, resolvedTools, resolvedSkills }) {
  const declared = applyPolicyToWorkflow(workflow, effective, { toolDomains: toolRegistry.TOOL_DOMAIN }) || workflow;
  const skills = Array.isArray(resolvedSkills)
    ? resolvedSkills.map(skillNameFromPath).filter(Boolean)
    : declared.skills || [];
  const tools = Array.isArray(resolvedTools)
    ? resolvedTools.map((tool) => tool && tool.name).filter(Boolean)
    : declared.tools || [];
  return { skills, tools, plugins: declared.mcp || [] };
}

/**
 * Run a workflow agent to completion on a single user message and return the
 * result plus the final assistant text. For repo-less workflows (no backend
 * provided) a scratch dir is created and cleaned up automatically.
 * @returns {Promise<{ result:object, messages:object[], finalText:string }>}
 */
async function runWorkflow({
  workflow,
  llm,
  userMessage,
  backend,
  skillPaths,
  rootDir,
  ctx = {},
  invokeConfig = {},
  runtime = 'deepagent',
  workflowPattern = 'sequential',
  env,
  rubric,
  rubricOptions,
  rubricMiddleware,
  settings,
  attribution,
  extraTools = [],
}) {
  // Resolve resource governance once, before any skill installation or MCP
  // connection. The same descriptor is then used for the build and trace
  // metadata so denied plugins cannot be loaded or reported as available.
  const effectiveWorkflow = applyPolicyToWorkflow(
    workflow,
    ctx.effectivePolicy || null,
    { toolDomains: toolRegistry.TOOL_DOMAIN },
  ) || workflow;
  let scratch = null;
  if (!backend && !rootDir) {
    scratch = prepareScratch(effectiveWorkflow);
    rootDir = scratch.rootDir;
    skillPaths = scratch.skillPaths;
  }
  try {
    const requestedRuntime = normalizeAgentRuntime(runtime, { strict: true });
    const runtimeId = effectiveAgentRuntime(requestedRuntime, llm, {
      strict: true,
      workflow: effectiveWorkflow.name,
      effectivePolicy: ctx.effectivePolicy || null,
    });
    const config = {
      recursionLimit: effectiveWorkflow.recursionLimit || 24,
      tags: effectiveWorkflow.tags || [],
      ...invokeConfig,
    };
    let deepAgentInvoke;
    let resolvedTools = null;
    let resolvedSkills = null;
    if (runtimeId === 'deepagent') {
      const mcpTools = await require('./mcp').loadMcpTools(effectiveWorkflow.mcp, ctx);
      const { agent, tools, skillPaths: builtSkills } = buildAgent({
        workflow: effectiveWorkflow,
        llm,
        backend,
        skillPaths,
        rootDir,
        ctx,
        extraTools: [...mcpTools, ...extraTools],
        env,
      });
      resolvedTools = tools;
      resolvedSkills = builtSkills;
      deepAgentInvoke = (prompt, tracedConfig) => {
        // The runtime wrapper owns the LangSmith root run id. Passing the same
        // id into the nested LangGraph invocation would create a duplicate run.
        const childConfig = { ...tracedConfig };
        delete childConfig.runId;
        return agent.invoke({ messages: [{ role: 'user', content: prompt }] }, childConfig);
      };
    }
    // Stamp the configured skills/tools/plugins onto the trace so runs are
    // filterable/summarisable by which resources were made available.
    const configWithResources = withResources(
      config,
      configuredResourceNames({
        workflow: effectiveWorkflow,
        effective: ctx.effectivePolicy || null,
        resolvedTools,
        resolvedSkills,
      })
    );
    return executeAgentRuntime({
      runtime: requestedRuntime,
      workflowPattern,
      prompt: userMessage,
      workflow: effectiveWorkflow.name,
      llm,
      rootDir,
      backendKind: effectiveWorkflow.backend,
      systemPrompt: effectiveWorkflow.systemPrompt,
      maxTurns: effectiveWorkflow.recursionLimit || 24,
      ctx,
      env,
      invokeConfig: configWithResources,
      tags: effectiveWorkflow.tags || [],
      deepAgentInvoke,
      lastText,
      // Optional RubricMiddleware completion review. Absent → no review runs.
      rubric,
      rubricOptions,
      rubricMiddleware,
      settings,
      // Billing attribution (see billing/usage.js). Prefer an explicit value;
      // else fall back to ctx.attribution so callers can thread it either way.
      attribution: attribution || ctx.attribution || null,
    });
  } finally {
    if (scratch) scratch.cleanup();
  }
}

/** Normalize message content (string or content-block array) to plain text. */
function contentToText(content) {
  if (typeof content === 'string') return content;
  if (Array.isArray(content)) return content.map((c) => (typeof c === 'string' ? c : c.text || '')).join('');
  return '';
}

/**
 * Text of a chat message, tolerant of REASONING models. Some models (e.g. LM
 * Studio's ornith) emit their answer as reasoning tokens and leave `content`
 * empty, putting the visible text in `additional_kwargs.reasoning_content`. When
 * `content` is blank we fall back to that so downstream parsing still sees the
 * answer, regardless of the JSON mode.
 */
function messageText(msg) {
  if (!msg) return '';
  const primary = contentToText(msg.content);
  if (primary && primary.trim()) return primary;
  const ak = msg.additional_kwargs || {};
  return contentToText(ak.reasoning_content || ak.reasoning || '');
}

function lastText(result) {
  const messages = (result && result.messages) || [];
  return messageText(messages[messages.length - 1]);
}

module.exports = {
  installSkills,
  buildBackend,
  loadWorkflow,
  prepareScratch,
  buildAgent,
  configuredResourceNames,
  runWorkflow,
  contentToText,
  messageText,
  lastText,
  SKILLS_DEST_DIRNAME,
};
